### ISSUE TITLE
Jobs are not enqueued by webhooks (GitHub)

### ISSUE THREAD

**Issue by morremeyer** [mend/renovate-ce-ee#357]
Running renovate-ce `6.1.1-full` with GitHub, jobs are not enqueued when webhooks are processed.

From the log:

```
{"name":"renovate","hostname":"renovate-mend-renovate-ce-789bb8cfff-t8vbt","pid":8,"level":20,"logContext":"fp7nLaycvNEuo36DM2hFV","event":{"organizationName":"redacted","installationId":123456789,"repositories":["redacted/repo"],"action":"enqueue","reason":"dependency-dashboard"},"msg":"Processed GitHub webhook event","time":"2023-10-05T13:00:56.563Z","v":0}
```

Querying the admin API directly afterwards:

```
curl -s -H "Authorization: PASSWORD" localhost:8080/api/job/queue  | jq -r '.[] | select(.repository == "redacted/repo")'
```

shows no output. 

Checking the first and last entries (since the queue always contains 100 items) shows:

```
curl -s -H "Authorization: PASSWORD" localhost:8080/api/job/queue  | jq -r '.[0:2], .[97:99]'
[
  {
    "jobId": 7,
    "repository": "redacted-one/repo-one",
    "organizationName": "redacted-one",
    "installationId": 0123456789,
    "reason": "repositories-added",
    "priority": 100,
    "addedAt": "2023-10-05 12:57:23"
  },
  {
    "jobId": 8,
    "repository": "redacted-one/repo-two",
    "organizationName": "redacted-one",
    "installationId": 0123456789,
    "reason": "repositories-added",
    "priority": 100,
    "addedAt": "2023-10-05 12:57:23"
  }
]
[
  {
    "jobId": 104,
    "repository": "redacted-two/repo-one",
    "organizationName": "redacted-two",
    "installationId": 987654321,
    "reason": "repositories-added",
    "priority": 100,
    "addedAt": "2023-10-05 12:57:24"
  },
  {
    "jobId": 105,
    "repository": "redacted-two/repo-two",
    "organizationName": "redacted-two",
    "installationId": 987654321,
    "reason": "repositories-added",
    "priority": 100,
    "addedAt": "2023-10-05 12:57:24"
  }
]
```

Manually enqueuing via the Admin API (`POST /api/job/add`) also does not add it to the job queue.

**Comment by PhilipAbed** [comment]
Hi, How are you running?

**Comment by morremeyer** [comment]
@PhilipAbed Do you mean how we're deploying renovate-ce? We're using the helm chart with a GitHub App.

**Comment by Gabriel-Ladzaretti** [comment]
The job/queue api returns the _first 100_ highest priority jobs from the queue. The queue size itself is unlimited, therefore the job might have been queued. Could you either inspect the sqlite db directly or try again when the queue is empty/smaller than 100 jobs and see if the jon is indeed enqueued?

**Comment by Gabriel-Ladzaretti** [comment]
I think we should tweak the job priory so that api requested jobs have higher priority than 'regular' requested ones e.g. webhooks.

**Comment by morremeyer** [comment]
I can check out the DB, but just to clarify the expectation that I have without having documentation available:

- The job queue is an array, I'd expect the first element to be the next job to run next
- If that's not the case, I'd expect the element with the lowest priority to run next. However, I can't see that since I can only see 100 items, making me fall back to the first assumption.

In line with the first assumption, I would expect lower (= more important) priority jobs.

Following the priorities listed in https://github.com/mend/renovate-ce-ee/blob/2305fc404bb29ca9e2ca771be9ef558321316a9d/docs/advanced.md#worker, I would expect jobs triggered from a webhook in the dependency dashboard to have a priority > 20 and < 40, although this is not documented anywhere so far (a quick check on our renovate on-prem instance didn't provide useful insight).

@Gabriel-Ladzaretti I agree with your idea that API requested should have a high priority, given the current docs, here's what I would do:

```
  ('onboarding-update', 20),
--> ('api-requested', 25),
--> ('dependency-dashboard', 26),
  ('pr-update', 30),
  ('closed-pr-rename', 35),
  ('manual-pr-merge', 40),
  ('repositories-added', 50),
  ('installed', 60),
  ('master-commit', 70),
  ('automerge', 80),
  ('manual-pr-close', 90),
  ('scheduled', 100);
```

In the current state with `6.1.1`, it looks like webhook requested runs are lower priority than `repository-added` ones, which at least to me is wrong - the purpose of having the possibility to manually request a run always struck me as "do it asap so users have quick feedback".

**Comment by Gabriel-Ladzaretti** [comment]
Unfortunately this table is outdated, in the current state, higher priority (numeric) means it will run first. 
requested events will be scanned before scheduled. repository-added is webhook based corresponds to a repository being added by the user in the app configuration page (github).

your 1st assumptions is correct, the job/queue api returns the head of the queue.

**Comment by Gabriel-Ladzaretti** [comment]
Note that jobs with the same priority will be resolved based on their age, the oldest will be scanned first.

**Comment by morremeyer** [comment]
That might explain why webhooks are not prioritizing requests: When the AppSync runs, it adds all repositories and the reason shows as `repositories-added` as seen above.

If the webhook also has the `repositories-added` reason, it stands to reason that webhook requested runs will just be queued behind the AppSync added ones. 

Does that sound right?

Combined with the sqlite currently not being persisted by the helm chart by default, since the [volume is mounted to `/tmp/renovate`](https://github.com/mend/renovate-ce-ee/blob/e2710b0bd591f082e2842ce9ce0320faa2aae6bc/helm-charts/mend-renovate-ce/templates/deployment.yaml#L175-L176) and the [sqlite is in `/tmp/renovate-ce.sqlite`](https://github.com/mend/renovate-ce-ee/blob/e2710b0bd591f082e2842ce9ce0320faa2aae6bc/helm-charts/mend-renovate-ce/values.yaml#L39) by default, this behaviour should re-occur every time the pod is restarted.

Since we have a lot of repositories and during debugging, restarts are frequent, it is possible that it just ran on the same repositories all the time since the queue was **not persisted**.

I see three things to (potentially) change:

- Update the priorities documentation
- If applicable: decrease priority of repositories being added by the AppSync
- Ensure the sqlite is persisted by default, either by moving it to `/tmp/renovate/renovate-ce.sqlite` by default or by mounting the PVC at `/tmp` by default (would be a breaking change since the PVC contents would be invalid and needed to be cleaned up).

I can open a PR for persisting the sqlite DB, for the other changes maintainer insight and implementation is needed.

**Comment by rarkins** [comment]
@Gabriel-Ladzaretti it would be good if the _initial_ sync on startup treated jobs enqueued as a result as _scheduled_ priority, instead of repository-added

**Comment by morremeyer** [comment]
@Gabriel-Ladzaretti To answer your questions about the status code, I get a HTTP 200 back:

```
❯ http --print hHbB post http://localhost:8080/api/job/add Authorization:PASSWORD repository=org/repo
POST /api/job/add HTTP/1.1
Accept: application/json, */*;q=0.5
Accept-Encoding: gzip, deflate
Authorization: PASSWORD
Connection: keep-alive
Content-Length: 32
Content-Type: application/json
Host: localhost:8080
User-Agent: HTTPie/3.2.2

{
    "repository": "org/repo"
}


HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 2
Content-Type: text/plain; charset=utf-8
Date: Fri, 06 Oct 2023 09:22:01 GMT
ETag: W/"2-nOO9QiTIwXgNtWtBJezz8kv3SLc"
Keep-Alive: timeout=5
X-Powered-By: Express

OK
```

**Comment by Gabriel-Ladzaretti** [comment]
200 usually means the job was successfully queued
404 repo doesnt exists in the db
500 exception when queuing/querying the db

**Comment by morremeyer** [comment]
This is resolved with 6.3.0, I can confirm that.

The underlying issue was a combination of sqlite not being persisted by default, which led to all repos being enqueued on every start combined with the priority for jobs of newly discovered repositories being equally high (or higher) than for webhook-requested jobs.

All of that is fixed now.
